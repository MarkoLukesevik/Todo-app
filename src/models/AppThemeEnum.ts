export enum AppThemeEnum {
    Light = 0,
    Dark = 1
}
