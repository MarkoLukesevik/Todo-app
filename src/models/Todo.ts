export interface Todo {
  id: string
  title: string
  isComplete: false
}
