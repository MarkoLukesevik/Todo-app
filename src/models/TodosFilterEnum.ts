export enum TodosFilterEnum {
    All = 0,
    Active = 1,
    Completed = 2
}