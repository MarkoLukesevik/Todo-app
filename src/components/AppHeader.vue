<script setup lang="ts">
import { computed } from 'vue'
import { AppThemeEnum } from '@/models/AppThemeEnum'

const props = defineProps<{
  appTheme: AppThemeEnum
}>()

const emit = defineEmits<{
  (e: 'toggleTheme', value: AppThemeEnum): void
}>()

const themeIcon = computed<string>(() => {
  return props.appTheme === AppThemeEnum.Light
    ? '/assets/images/icon-moon.svg'
    : '/assets/images/icon-sun.svg'
})

const handleIconClick = (): void => {
  props.appTheme === AppThemeEnum.Light
    ? emit('toggleTheme', AppThemeEnum.Dark)
    : emit('toggleTheme', AppThemeEnum.Light)
}
</script>

<template>
  <div class="app-header">
    <h1 class="app-header__title">TODO</h1>
    <img class="app-header__theme-icon" :src="themeIcon" alt="" @click="handleIconClick" />
  </div>
</template>

<style scoped lang="scss">
.app-header {
  display: flex;
  flex-direction: row;
  flex: 0 1 auto;
  align-items: center;
  justify-content: space-between;

  color: #ffffff;

  font-family: 'Josefin Sans', sans-serif;

  &__title {
    font-size: 2rem;
  }
}
</style>
