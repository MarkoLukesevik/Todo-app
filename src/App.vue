<script setup lang="ts">
import { ref, computed } from 'vue';
import type { Ref } from 'vue';

import { AppThemeEnum } from '@/models/AppThemeEnum';

import TodosContainer from '@/components/TodosContainer/TodosContainer.vue';
import AppHeader from '@/components/AppHeader.vue';

const appTheme: Ref<AppThemeEnum> = ref(AppThemeEnum.Light);

const appBackgroundImage = computed<string>(() => {
    if (window.innerWidth < 768) {
      return appTheme.value === AppThemeEnum.Light
        ? '/assets/images/bg-mobile-light.jpg'
        : '/assets/images/bg-mobile-dark.jpg';
    } else {
      return appTheme.value === AppThemeEnum.Light
        ? '/assets/images/bg-desktop-light.jpg'
        : '/assets/images/bg-desktop-dark.jpg';
    }
})

const appBackgroundColor = computed<string>(() => {
    return appTheme.value === AppThemeEnum.Light ? '#fafafa' : '#171823';
});

const handleThemeToggle = (theme: AppThemeEnum): void => {
    appTheme.value = theme;
}
</script>

<template>
    <div class="app" :style="{ backgroundColor: appBackgroundColor }">
        <div :style="{ backgroundImage: 'url(' + appBackgroundImage + ')' }" class="app-image"></div>
        <div class="app-container">
            <app-header :appTheme="appTheme" @toggle-theme="handleThemeToggle" />
            <todos-container :app-theme="appTheme" />
        </div>
    </div>
</template>

<style scoped lang="scss">
@import url('https://fonts.googleapis.com/css2?family=Josefin+Sans:ital,wght@0,100..700;1,100..700&display=swap');
.app {
    position: relative;

    display: flex;
    flex: 1 1 auto;
    flex-direction: column;
    align-items: center;
    row-gap: 1.5rem;

    padding: 2rem 1.5rem;

    &-image {
        position: absolute;
        left: 0;
        right: 0;
        top: 0;

        background-size: cover;
        background-position: bottom right;
        background-repeat: no-repeat;

        width: 100%;
        height: 30%;
    }

    &-container {
        position: relative;

        display: flex;
        flex: 1 1 auto;
        flex-direction: column;
        row-gap: 1.5rem;

        width: 100%;
        margin: 0 auto;
    }
}

@media only screen and (min-width: 768px) {
.app {
    row-gap: 2rem;
    padding: 3rem 1.5rem;

    &-container {
        row-gap: 2rem;
        max-width: 540px
    }
}
}
</style>
