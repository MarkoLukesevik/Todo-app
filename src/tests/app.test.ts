import { shallowMount } from '@vue/test-utils';
import { describe, expect, it, beforeEach, afterEach } from 'vitest';
import { nextTick } from 'vue';

import { AppThemeEnum } from '@/models/AppThemeEnum';
import { mockLocalStorage } from './mockLocalStorage/mockLocalStorage';

import App from '@/App.vue';
import AppHeader from '@/components/AppHeader.vue';

Object.defineProperty(window, 'localStorage', { value: mockLocalStorage });

describe('Test suite for app component', () => {
    let wrapper: any;
    let originalInnerWidth: number;

    beforeEach(() => {
        originalInnerWidth = window.innerWidth;
        localStorage.clear();
        wrapper = shallowMount(App);
    });

    afterEach(() => {
        Object.defineProperty(window, 'innerWidth', { value: originalInnerWidth, writable: true });
      });

    it('Should match the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should initialize appTheme from localStorage', async () => {
        localStorage.setItem('appTheme', JSON.stringify(AppThemeEnum.Dark));
        wrapper = shallowMount(App);
        expect(wrapper.vm.appTheme).toBe(AppThemeEnum.Dark);
    });
    
    it('should default appTheme to Light if no localStorage entry', () => {
        localStorage.removeItem('appTheme');
        expect(wrapper.vm.appTheme).toBe(AppThemeEnum.Light);
    });

    it('should compute correct background image for light theme', async () => {
        const imageDiv = await wrapper.find('[data-test="app-image"]');

        expect(imageDiv.element.style.getPropertyValue('background-image'))
            .toBe('url(/assets/images/bg-desktop-light.jpg)');
        expect(wrapper.vm.appBackgroundImage).toBe('/assets/images/bg-desktop-light.jpg');
            
        Object.defineProperty(window, 'innerWidth', { value: 200, writable: true });
        window.dispatchEvent(new Event('resize'));
       
        expect(wrapper.vm.appBackgroundImage).toBe('/assets/images/bg-mobile-light.jpg');
    });

    it('should compute correct background image for dark theme', async () => {
        const imageDiv = await wrapper.find('[data-test="app-image"]');
        wrapper.vm.appTheme = AppThemeEnum.Dark;
        await nextTick();

        expect(imageDiv.element.style.getPropertyValue('background-image'))
            .toBe('url(/assets/images/bg-desktop-dark.jpg)');
        expect(wrapper.vm.appBackgroundImage).toBe('/assets/images/bg-desktop-dark.jpg');
            
        Object.defineProperty(window, 'innerWidth', { value: 200, writable: true });
        window.dispatchEvent(new Event('resize'));
       
        expect(wrapper.vm.appBackgroundImage).toBe('/assets/images/bg-mobile-dark.jpg');
    });

    it('should compute and render correct app classes for dark theme', async () => {
        const appDiv = await wrapper.find('[data-test="app"]');
        wrapper.vm.appTheme = AppThemeEnum.Dark;
        await nextTick();

        expect(wrapper.vm.appClasses).toContain('app-dark');
        expect(appDiv.classes()).toEqual(['app', 'app-dark']);
    });
    
    it('should compute and render correct app classes for light theme', async () => {
        const appDiv = await wrapper.find('[data-test="app"]');
        wrapper.vm.appTheme = AppThemeEnum.Light;

        expect(wrapper.vm.appClasses).not.toContain('app-dark');
        expect(appDiv.classes()).toEqual(['app']);
    });

    it('should update appTheme and localStorage on theme toggle', async () => {
        expect(wrapper.vm.appTheme).toBe(AppThemeEnum.Light);

        wrapper.vm.handleThemeToggle(AppThemeEnum.Dark);
        await nextTick();

        expect(wrapper.vm.appTheme).toBe(AppThemeEnum.Dark);
        expect(localStorage.getItem('appTheme')).toBe(JSON.stringify(AppThemeEnum.Dark));
      });
    
      it('should handle theme toggle event from AppHeader', async () => {
        const appHeader = wrapper.findComponent(AppHeader);

        await appHeader.vm.$emit('toggle-theme', AppThemeEnum.Dark);
        await nextTick();

        expect(wrapper.vm.appTheme).toBe(AppThemeEnum.Dark);
      });
});