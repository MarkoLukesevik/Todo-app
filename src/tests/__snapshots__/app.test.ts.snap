// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test suite for app component > Should match the snapshot 1`] = `
<div
  class="app"
  data-test="app"
  data-v-7a7a37b1=""
>
  <div
    class="app-image"
    data-test="app-image"
    data-v-7a7a37b1=""
    style="background-image: url(/assets/images/bg-desktop-light.jpg);"
  />
  <div
    class="app-container"
    data-v-7a7a37b1=""
  >
    <app-header-stub
      apptheme="0"
      data-v-7a7a37b1=""
    />
    <todos-container-stub
      apptheme="0"
      data-v-7a7a37b1=""
    />
  </div>
</div>
`;

exports[`test suite for app component > Should match the snapshot 1`] = `
<div
  class="app"
  data-test="app"
  data-v-7a7a37b1=""
>
  <div
    class="app-image"
    data-test="app-image"
    data-v-7a7a37b1=""
    style="background-image: url(/assets/images/bg-desktop-light.jpg);"
  />
  <div
    class="app-container"
    data-v-7a7a37b1=""
  >
    <app-header-stub
      apptheme="0"
      data-v-7a7a37b1=""
    />
    <todos-container-stub
      apptheme="0"
      data-v-7a7a37b1=""
    />
  </div>
</div>
`;
